obtuse and confusing error with local feature under WSL #834

prathameshzarkar9 opened the pull request:
[Issue 834](https://github.com/devcontainers/cli/pull/838)
Confusing error message modified to more understandable message.

samruddhikhandale commented on src/spec-configuration/containerFeaturesConfiguration.ts:
```diff
@@ -698,8 +698,8 @@ export async function getFeatureIdType(params: CommonParams, userFeatureId: stri
 	if (!userFeatureId.includes('/') && !userFeatureId.includes('\\')) {
-		output.write(`Legacy feature '${userFeatureId}' not supported. Please check https://containers.dev/features for replacements.`, LogLevel.Error);
-		throw new ContainerError({ description: `Legacy feature '${userFeatureId}' not supported. Please check https://containers.dev/features for replacements.` });
+		output.write(`Legacy feature '${userFeatureId}' forgot to prepend './' at the begining of the feature. Please check https://containers.dev/implementors/features-distribution/#addendum-locally-referenced for replacements.`, LogLevel.Error);
```
Can you follow the messaging in https://github.com/devcontainers/cli/issues/834#issuecomment-2148074401? Thanks! 

We need to append the new message and not replace it. The issue comment has the needed message string.

chrmarti commented on src/spec-configuration/containerFeaturesConfiguration.ts:
```diff
@@ -698,8 +698,11 @@ export async function getFeatureIdType(params: CommonParams, userFeatureId: stri
 	if (!userFeatureId.includes('/') && !userFeatureId.includes('\\')) {
-		output.write(`Legacy feature '${userFeatureId}' not supported. Please check https://containers.dev/features for replacements.`, LogLevel.Error);
-		throw new ContainerError({ description: `Legacy feature '${userFeatureId}' not supported. Please check https://containers.dev/features for replacements.` });
+		output.write(`Legacy feature '${userFeatureId}' not supported. Please check https://containers.dev/features for replacements. \n
```
- Since it is twice the same message, could you extract it to a local variable?
- There is indentation on the second line of the message that needs to be removed.
- You could use an actual newline in stead of `\n`.

prathameshzarkar9 replied:
@samruddhikhandale @chrmarti,
kindly check the modifications done to address this.

chrmarti reviewed (changes requested):
Thanks for the PR, left a comment.

prathameshzarkar9 commented:
Hi @samruddhikhandale ,
I have addressed review comments and also modified the error message as suggested by @chrmarti .
Kindly review and approve the same.

samruddhikhandale commented on src/spec-configuration/containerFeaturesConfiguration.ts:
```diff
@@ -686,6 +686,8 @@ export function updateDeprecatedFeaturesIntoOptions(userFeatures: DevContainerFe
 
 export async function getFeatureIdType(params: CommonParams, userFeatureId: string, lockfile: Lockfile | undefined) {
 	const { output } = params;
+	const error_message = `Legacy feature '${userFeatureId}' not supported. Please check https://containers.dev/features for replacements.
```
```suggestion
	const errorMessage = `Legacy feature '${userFeatureId}' not supported. Please check https://containers.dev/features for replacements.
```

prathameshzarkar9 replied:
Done

samruddhikhandale commented on src/spec-configuration/containerFeaturesConfiguration.ts:
```diff
@@ -686,6 +686,8 @@ export function updateDeprecatedFeaturesIntoOptions(userFeatures: DevContainerFe
 
 export async function getFeatureIdType(params: CommonParams, userFeatureId: string, lockfile: Lockfile | undefined) {
 	const { output } = params;
+	const error_message = `Legacy feature '${userFeatureId}' not supported. Please check https://containers.dev/features for replacements.
```
Can we move this variable to L703 ie inside [this](https://github.com/devcontainers/cli/pull/838/files#diff-93622e1db0f140d14a85a65ca0a4522090b385c3964de02011229c0523095470R702) condition?

prathameshzarkar9 replied:
done

samruddhikhandale commented on src/spec-configuration/containerFeaturesConfiguration.ts:
```diff
@@ -686,6 +686,8 @@ export function updateDeprecatedFeaturesIntoOptions(userFeatures: DevContainerFe
 export async function getFeatureIdType(params: CommonParams, userFeatureId: string, lockfile: Lockfile | undefined) {
 	const { output } = params;
+	const error_message = `Legacy feature '${userFeatureId}' not supported. Please check https://containers.dev/features for replacements.
+	If you were hoping to use local Features, remember to prepend your Feature name with "./". Please check https://containers.dev/implementors/features-distribution/#addendum-locally-referenced for more information.`;
```
```suggestion
If you were hoping to use local Features, remember to prepend your Feature name with "./". Please check https://containers.dev/implementors/features-distribution/#addendum-locally-referenced for more information.`;
```

Should we remove the tab?

prathameshzarkar9 replied:
tab removed

samruddhikhandale commented on src/test/cli.build.test.ts:
```diff
@@ -27,6 +27,17 @@ describe('Dev Containers CLI', function () {
+			} catch (error) {
+				const res = JSON.parse(error.stdout);
+				assert.equal(res.outcome, 'error');
+				assert.match(res.message, /prepend your Feature name with/);
```
Thanks for adding the test!

samruddhikhandale commented on src/test/cli.build.test.ts:
```diff
@@ -27,6 +27,17 @@ describe('Dev Containers CLI', function () {
+			} catch (error) {
+				const res = JSON.parse(error.stdout);
+				assert.equal(res.outcome, 'error');
+				assert.match(res.message, /prepend your Feature name with/);
```
Nit: We can also assert for the entire error message.

prathameshzarkar9 replied:
error message contains the dynamic feature name which is defined in the `devcontainer.json` file. Referred as `userFeatureId` where the error message is modified.
So used the match.

samruddhikhandale reviewed (approved):
Looks good to me, let's wait for another set of review from @chrmarti 

thanks!

chrmarti reviewed (approved):
LGTM, thanks!
